### DISCUSSION TITLE
The issue of protocol implementation.

### DISCUSSION THREAD

**Discussion by R1BGW** [semuconsulting/PyGPSClient#150] [General]
Is there an implementation of data exchange via the garmin binary protocol (L000, L001,D600) and others?

**Comment by semuadmin** [comment]
No, sorry. There are no current plans to implement Garmin’s proprietary binary protocol, though it’s a possibility for the future assuming the specifications are in the public domain. The ‘pynmeagps’ library does support several proprietary Garmin NMEA sentences. 

 Contributions always welcome.

**Reply by R1BGW** [reply]
There is a specification , and users also have a need , for example: a line of Garmin Striker echo sounders .
[garmin-binary.pdf](https://github.com/user-attachments/files/16797194/garmin-binary.pdf)

**Reply by R1BGW** [reply]
Thanks anyway.

**Reply by semuadmin** [reply]
> There is a specification , and users also have a need , for example: a line of Garmin Striker echo sounders . [garmin-binary.pdf](https://github.com/user-attachments/files/16797194/garmin-binary.pdf)

I don’t doubt there’s a need, but part of the problem is that AFAIK no two Garmin receivers implement exactly the same protocol, so development and testing would involve acquiring multiple Garmin devices to ensure reasonable coverage. 

Please appreciate that PyGPSClient and its component libraries are maintained entirely by unpaid volunteers - we receive absolutely no corporate or advertising funding - so our resources are limited without contributions.

**Reply by R1BGW** [reply]
I understand perfectly well. I tried to implement a program based on the PyGarmin library, I even managed to connect 2 different receivers via the serial port (COM) and unload points and tracks from them, as well as turn them on in PVT mode (GPSmap60CSx, Garmin Sriker 4+) from the command line. I ran into an incorrect GPX format output and the inability to connect via USB (gives an error) . These two points baffled me. I am not a programmer by profession, although I work with electronics.
[get-tracks (1).txt](https://github.com/user-attachments/files/16797598/get-tracks.1.txt)
[get-waypoints.txt](https://github.com/user-attachments/files/16797599/get-waypoints.txt)
[gfhjkm.txt](https://github.com/user-attachments/files/16797600/gfhjkm.txt)
[info.txt](https://github.com/user-attachments/files/16797601/info.txt)
[pvt.txt](https://github.com/user-attachments/files/16797602/pvt.txt)
Examples

**Reply by semuadmin** [reply]
Aside from the cost of the development hardware (_UBX devices can be picked up for $10 or less - Garmin devices are generally $100s_), the problem - as I say - is that there is no one binary Garmin protocol for **real-time navigational data** across all Garmin devices. There is a certain degree of commonality between product verticals (maritime, aviation, automotive, consumer, etc.) but still significant variability between individual models. In addition, much of the [Garmin SDK](https://www8.garmin.com/support/commProtocol.html) (software development kit) is concerned with the offline transfer of routes, tracks, waypoints and almanac data between proprietary Garmin devices, which is not really what PyGPSClient and its underlying libraries are all about (_our interest is primarily real time navigation and RTK data and device configuration_). The SDK references the D800 PVT protocol for real time navigation data  - e.g. example below - but as far as I can see there's nothing here that isn't available via the standard or proprietary NMEA sentences already supported. In terms of Garmin device configuration, each implementation is more or less device-specific.

```
7.5.31. D800_Pvt_Data_Type
Example products: GPS III and StreetPilot.

typedef struct
{
float alt; /* altitude above WGS 84 ellipsoid (meters) */
float epe; /* estimated position error, 2 sigma (meters) */
float eph; /* epe, but horizontal only (meters) */
float epv; /* epe, but vertical only (meters) */
int fix; /* type of position fix */
double tow; /* time of week (seconds) */
Radian_Type posn; /* latitude and longitude (radians) */
float east; /* velocity east (meters/second) */
float north; /* velocity north (meters/second) */
float up; /* velocity up (meters/second) */
float msl_hght; /* height of WGS 84 ellipsoid above MSL (meters) */
int leap_scnds; /* difference between GPS and UTC (seconds) */
long wn_days; /* week number days */
} D800_Pvt_Data_Type;
```

Out of interest, what _specific capabilities_ were you interested in being added to PyGPSClient that aren't already supported? Happy to explore further if there's significant value to be added.

**Reply by R1BGW** [reply]
In general, I would like to be able to create maps on my own (for reservoirs, drawing depths from tracks , for the earth's surface, drawing height contours). The ability to add navigation signs . There are analogs of programs, but they are more or less not so flexible. The main exchange format is used by GPX, but for example, the ReefMaster program uses the CSV extension to build underwater maps. For the GPSMapEdit program (can draw maps of the earth's surface but cannot draw depths)GPX or CSV is enough . I would like to see it as a mobile tool for working with the ability to add maps not only online but also upload your own .

**Reply by gdt** [reply]
Sounds like fun stuff to do but it doesn't really seem to fit this project.  Besides a garmin project to get data out, you probably want to use qgis to manage your data and also look at mkgmap (from the openstreetmap world, top DDG hit seems like what you want) that can take map input data (in vector form) and produce garmin-compatible .img files.   I use that to get OSM data on a nuvi and an etrex 30.     Not about this program, so I'll give that hint and not follow up here.

**Reply by semuadmin** [reply]
> Sounds like fun stuff to do but it doesn't really seem to fit this project. Besides a garmin project to get data out, you probably want to use qgis to manage your data and also look at mkgmap (from the openstreetmap world, top DDG hit seems like what you want) that can take map input data (in vector form) and produce garmin-compatible .img files. I use that to get OSM data on a nuvi and an etrex 30. Not about this program, so I'll give that hint and not follow up here.

@R1BGW @gdt 
Thanks to both for sharing - I was about to make exactly the same recommendation myself. Assuming the apps that are bundled with Garmin devices don't fit the bill, [QGIS](https://qgis.org/) is a very powerful free, multi-platform open-source GIS (_geographical information system_) application which I use extensively myself. Assuming you can source a suitable maritime basemap (_or at least a set of geo-referenced images_), QGIS can create waypoints, tracks, polygons and all manner of geographical features. Its core functionality can be extended via custom Python modules and there's a very active developer community. It does carry something of a steep learning curve, but it tends to reward the effort.

[mkgmap](https://www.mkgmap.org.uk/) is essentially a geo-referencing utility - it converts [Open Street Map (OSM)](https://www.openstreetmap.org/#map=6/54.91/-3.43) data into geo-referenced *.img files which can be uploaded into compatible Garmin devices. I'm not sure if it can be extended to use other map sources, but I'm sure this'll be covered in the documentation. Features are being added to QGIS all the time, so its possible that at least some of its capabilities are now core QGIS functions.

The downloading of tracks and waypoints from Garmin receivers is not really an intended application for PyGPSClient. PyGPSClient offers a basic facility to record GPX tracks from raw GNSS datastreams (_and I have a request in my backlog to enable its [GPX viewer](https://github.com/semuconsulting/PyGPSClient?tab=readme-ov-file#gpxviewer) to use custom offline basemaps rather than the MapQuest API, which I'll get to at some point_), but it's not really intended to be used to create map content.

**Reply by semuadmin** [reply]
FYI if you want to display your own custom maps in PyGPSClient, there is a [Python script `georef.py`](https://github.com/semuconsulting/PyGPSClient/blob/master/examples/georef.py) in the /examples folder which helps you do this.
